README.md
# Tokenization Service Contracts

The Tokenization Service's Smart Contract.

## How to use the smart contracts:

1. Install the tokenization service contracts:

```
npm install @jwalab/tokenization-service-contracts
```

2. Deploy with Taquito:

Example deploying the warehouse contract:

```
const { TezosToolkit, MichelsonMap } = require('@taquito/taquito');
const { InMemorySigner } = require('@taquito/signer');

const tokenizationServiceContracts = require('@jwalab/tokenization-service-contracts');

// using truffle's default accounts
// An account looks like this:
// alice: {
//      pkh: "tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb",
//      sk: "edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq",
//      pk: "edpkvGfYw3LyB1UcCahKQk4rF2tvbMUk8GFiTuMjL75uGXrpvKXhjn"
// }
const accounts = require('./path/to/accounts');

const Tezos = new TezosToolkit();

Tezos.setProvider({
    rpc: 'http://localhost:20000',
    signer: new InMemorySigner(accounts.alice.sk)
})

Tezos.contract
  .originate({
    code: tokenizationServiceContracts.warehouse.michelson,
    storage: {
        owners: ["tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb"],
        version: "1",
        items: MichelsonMap.fromLiteral({}),
        instances: MichelsonMap.fromLiteral({})
    },
  })
  .then((originationOp) => {
    console.log(`Waiting for confirmation of origination for ${originationOp.contractAddress}`);
    return originationOp.contract(1, 1);
  })
  .then(() => {
    console.log(`Origination completed.`);
  })
  .catch((error) => console.log(error));
```

### Smart Contract APIs

Our Smart Contracts are fully tested, please look at the `./test` folder to see how they work and how to configure their storage.

## Development

### Requirements

1. Items created in the Warehouse are truly immutable, their characteristics may never be altered. Just like a manufactured product once manufactured can't be altered. If a faulty item is minted, it must be discarded or sold as-is, or minted again.
2. Items in the Warehouse don't belong to anyone and can't be transferred. Items in the Warehouse are not owned, only linked to an originator.
3. Items in the Warehouse are semi-fungible. If an item is created with a quantity of 1000, all 1000 items have the same value and can be exchanged without destruction of value.
4. When an item in the Warehouse is purchased, and item instance is created and assigned to a user. The Warehouse item remains unaltered, its total quantity remains unaltered, but there's one fewer item left to be purchased (available_quantity). The instance is now mutable and can evolve independently from other instances of the same item.
5. An item's instance can be transferred to another user.
6. One more thing, an item in the Warehouse can actually be modified until the it's marked as `frozen`.

### Project Structure

The Tezos contracts are written in the `./contracts` folder.

### Getting Started

1. Start by cloning this repository
1. run `npm install` in this folder

#### Test the contracts using the built-in sandbox

Start a tezos node first, use minilab or

```
docker run -d -p 20000:20000 -e block_time=2 tqtezos/flextesa:20210602 granabox start
```

Then run the tests:

```
npm run test
```

### Cost estimation:

#### 0.2.0 with multiple owners

| action                             | cost        |
| ---------------------------------- | ----------- |
| warehouse origination              | `0.626894ꜩ` |
| create warehouse item              | `0.031207ꜩ` |
| Update same size warehouse item    | `0.000733ꜩ` |
| Update smaller size warehouse item | `0.000708ꜩ` |
| Update bigger size warehouse item  | `0.017005ꜩ` |
| Freeze item                        | `0.000680ꜩ` |
| Assign item to user                | `0.021997ꜩ` |
| Update instance                    | `0.009729ꜩ` |
| transfer instance to new user      | `0.000703ꜩ` |

#### 0.1.1 with security

| action                             | cost        |
| ---------------------------------- | ----------- |
| warehouse origination              | `0.566903ꜩ` |
| create warehouse item              | `0.031202ꜩ` |
| Update same size warehouse item    | `0.000729ꜩ` |
| Update smaller size warehouse item | `0.000703ꜩ` |
| Update bigger size warehouse item  | `0.017045ꜩ` |
| Freeze item                        | `0.000675ꜩ` |
| Assign item to user                | `0.021992ꜩ` |
| Update instance                    | `0.009724ꜩ` |
| transfer instance to new user      | `0.000698ꜩ` |

#### 0.1.0 integrated contract and Granada:

| action                             | cost        |
| ---------------------------------- | ----------- |
| warehouse origination              | `0.57152ꜩ`  |
| create warehouse item              | `0.032072ꜩ` |
| Update same size warehouse item    | `0.001618ꜩ` |
| Update smaller size warehouse item | `0.001588ꜩ` |
| Update bigger size warehouse item  | `0.017936ꜩ` |
| Freeze item                        | `0.001552ꜩ` |
| Assign item to user                | `0.022876ꜩ` |
| Update instance                    | `0.010594ꜩ` |
| transfer instance to new user      | `0.001568ꜩ` |

#### Granada:

| action                             | cost        |
| ---------------------------------- | ----------- |
| warehouse origination              | `0.451211ꜩ` |
| inventory origination              | `0.386428ꜩ` |
| create warehouse item              | `0.031179ꜩ` |
| Update same size warehouse item    | `0.000706ꜩ` |
| Update smaller size warehouse item | `0.00068ꜩ`  |
| Update bigger size warehouse item  | `0.017023ꜩ` |
| Transfer item to inventory         | `0.020832ꜩ` |

#### Florence:

| action                             | cost        |
| ---------------------------------- | ----------- |
| warehouse origination              | `0.451804ꜩ` |
| inventory origination              | `0.386914ꜩ` |
| create warehouse item              | `0.03179ꜩ`  |
| Update same size warehouse item    | `0.001335ꜩ` |
| Update smaller size warehouse item | `0.001305ꜩ` |
| Update bigger size warehouse item  | `0.017653ꜩ` |
| Transfer item to inventory         | `0.022413ꜩ` |


CHANGELOG.md
# Changelog

## 0.2.0

-   Breaking change! The contract storage is initialized with a set of addresses instead of a single address
-   It's now possible to have multiple owners for the storage

## 0.1.1

-   An item doesn't need to be frozen before it can be assigned. A user may be assigned an item that can still change
    but it's an agreement between an item editor and the user.
-   Ensure that the contract owner is the only user allowed to execute the contract for security reasons.

## 0.1.0

-   Complete overhaul of the contracts. Warehouse and Inventory are now merged
    into one contract with 2 bigmaps. Instances don't live in a new contract but in a collocated bigmap.
    The goal was to reduce the cost of operation and the code's complexity.
-   Update cost estimation script
-   Upgrade to Taquito 10

## 0.0.14

-   Add missing type for transfer_item method

## 0.0.13

-   Fix npm package entrypoint

## 0.0.12

-   Add transfer_item entrypoint to inventory

## 0.0.11

-   No changes in the contracts
-   Removed truffle for testing and deploying
-   Remove migrations scripts and contract
-   Upgrade to Florence
-   Add scripts to estimate cost of standard operations


tests/utils.js
async function getWarehouseInstanceAt(storage, itemId, instanceNumber) {
    const instanceObject = await storage.instances.get({
        item_id: itemId,
        instance_number: instanceNumber
    });

    return warehouseInstanceToObject(instanceObject);
}

async function hasWarehouseInstanceAt(storage, itemId, instanceNumber) {
    const instanceObject = await storage.instances.get({
        item_id: itemId,
        instance_number: instanceNumber
    });

    return Boolean(instanceObject);
}

function warehouseInstanceToObject(michelsonObject) {
    return {
        user_id: michelsonObject.user_id,
        data: Object.fromEntries(michelsonObject.data.entries())
    };
}

function warehouseItemToObject(item) {
    return {
        available_quantity: item.available_quantity.toNumber(),
        data: Object.fromEntries(item.data.entries()),
        frozen: item.frozen,
        item_id: item.item_id.toNumber(),
        name: item.name,
        total_quantity: item.total_quantity.toNumber()
    };
}

async function originateContract(tezos, code, storage) {
    const originatonOperation = await tezos.contract.originate({
        code,
        storage
    });

    const { contractAddress } = originatonOperation;

    await originatonOperation.confirmation(1, 1);

    return tezos.contract.at(contractAddress);
}

module.exports = {
    warehouseItemToObject,
    warehouseInstanceToObject,
    getWarehouseInstanceAt,
    hasWarehouseInstanceAt,
    originateContract
};


tests/owners.test.js
const { TezosToolkit, MichelsonMap } = require("@taquito/taquito");
const { InMemorySigner } = require("@taquito/signer");

const { originateContract } = require("./utils");

const warehouseContract = require("../build/contracts/warehouse.json");

describe("Given Warehouse is deployed", () => {
    let warehouseInstance;
    let tezos;

    beforeAll(async () => {
        tezos = new TezosToolkit("http://localhost:20000");

        tezos.setProvider({
            signer: new InMemorySigner(
                "edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq"
            )
        });

        warehouseInstance = await originateContract(tezos, warehouseContract, {
            owners: ["tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb"],
            version: "1",
            items: MichelsonMap.fromLiteral({}),
            instances: MichelsonMap.fromLiteral({})
        });
    });

    describe("When removing the only user", () => {
        it("Then fails with an error as at least 1 owner is required", async () => {
            try {
                await warehouseInstance.methods
                    .remove_owner("tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb")
                    .send();

                throw new Error(
                    "Remove_owner should throw an error if the resulting list of owners is empty"
                );
            } catch (err) {
                expect(err.message).toEqual("REQUIRES_AT_LEAST_ONE_OWNER");
            }
        });
    });

    describe("When adding another user", () => {
        let owners;

        beforeEach(async () => {
            const operation = await warehouseInstance.methods
                .add_owner("tz1aSkwEot3L2kmUvcoxzjMomb9mvBNuzFK6")
                .send();

            await operation.confirmation(1);

            owners = (await warehouseInstance.storage()).owners;
        });

        it("Then adds the owner to the list", async () => {
            expect(owners).toEqual([
                "tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb",
                "tz1aSkwEot3L2kmUvcoxzjMomb9mvBNuzFK6"
            ]);
        });

        describe("When removing a user", () => {
            let owners;

            beforeEach(async () => {
                const operation = await warehouseInstance.methods
                    .remove_owner("tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb")
                    .send();

                await operation.confirmation(1);

                owners = (await warehouseInstance.storage()).owners;
            });

            it("Then removes the owner from the list", async () => {
                expect(owners).toEqual([
                    "tz1aSkwEot3L2kmUvcoxzjMomb9mvBNuzFK6"
                ]);
            });
        });
    });
});


tests/inventory.test.js
const { TezosToolkit, MichelsonMap } = require("@taquito/taquito");
const { InMemorySigner } = require("@taquito/signer");

const { originateContract, getInventoryItemAt } = require("./utils");

const inventoryContract = require("../build/contracts/inventory.json");

describe("Given Inventory is deployed", () => {
    let inventoryInstance;
    let storage;
    let tezos;

    beforeAll(async () => {
        tezos = new TezosToolkit("http://localhost:20000");

        tezos.setProvider({
            signer: new InMemorySigner(
                "edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq"
            )
        });

        inventoryInstance = await originateContract(
            tezos,
            inventoryContract,
            MichelsonMap.fromLiteral({})
        );
    });

    describe("When assigning a new item", () => {
        beforeAll(async () => {
            const operation = await inventoryInstance.methods
                .assign_item(
                    MichelsonMap.fromLiteral({
                        XP: "97"
                    }),
                    12,
                    2,
                    "user_123"
                )
                .send();

            await operation.confirmation(1);

            storage = await inventoryInstance.storage();
        });

        it("Then assigns the item to the user", async () => {
            const obj = await getInventoryItemAt(storage, "user_123", 2, 12);

            expect(obj).toEqual({
                data: {
                    XP: "97"
                }
            });
        });

        describe("When I update this instance", () => {
            beforeAll(async () => {
                const operation = await inventoryInstance.methods
                .update_item(
                    MichelsonMap.fromLiteral({
                        XP: "98"
                    }),
                    12,
                    2,
                    "user_123"
                )
                .send();

                await operation.confirmation(1);

                storage = await inventoryInstance.storage();
            });

            it("Then updates the instance", async () => {
                const obj = await getInventoryItemAt(storage, "user_123", 2, 12);

                expect(obj).toEqual({
                    data: {
                        XP: "98"
                    }
                });
            });
        });

        describe("And When I assign it again", () => {
            it("Then fails with an error", async () => {
                try {
                    const operation = await inventoryInstance.methods
                    .assign_item(
                        MichelsonMap.fromLiteral({
                            XP: "97"
                        }),
                        12,
                        2,
                        "user_123"
                    )
                    .send();

                    await operation.confirmation(1);

                    storage = await inventoryInstance.storage();

                    console.error(
                        "Will fail: Assign_Item should throw an Error if the item and instance have already been assigned"
                    );
    
                    fail(
                        "Assign_Item should throw an Error if the item and instance have already been assigned"
                    );
                } catch (err) {
                    expect(err.message).toBe("ITEM_INSTANCE_ALREADY_ASSIGNED");
                }
            });
        });

        describe("When I assign another instance of the same item", () => {
            beforeAll(async () => {
                const operation = await inventoryInstance.methods
                    .assign_item(
                        MichelsonMap.fromLiteral({
                            XP: "97"
                        }),
                        13,
                        2,
                        "user_123"
                    )
                    .send();
    
                await operation.confirmation(1);
    
                storage = await inventoryInstance.storage();
            });
    
            it("Then assigns the item to the user", async () => {
                const obj = await getInventoryItemAt(storage, "user_123", 2, 13);
    
                expect(obj).toEqual({
                    data: {
                        XP: "97"
                    }
                });
            });

            describe("When I update this instance", () => {
                beforeAll(async () => {
                    const operation = await inventoryInstance.methods
                    .update_item(
                        MichelsonMap.fromLiteral({
                            XP: "99"
                        }),
                        13,
                        2,
                        "user_123"
                    )
                    .send();
    
                    await operation.confirmation(1);
    
                    storage = await inventoryInstance.storage();
                });
    
                it("Then updates the instance", async () => {
                    const obj = await getInventoryItemAt(storage, "user_123", 2, 13);
    
                    expect(obj).toEqual({
                        data: {
                            XP: "99"
                        }
                    });
                });
            });
        });

        describe("When I assign another item to the same user", () => {
            beforeAll(async () => {
                const operation = await inventoryInstance.methods
                    .assign_item(
                        MichelsonMap.fromLiteral({
                            XP: "97"
                        }),
                        1,
                        3,
                        "user_123"
                    )
                    .send();
    
                await operation.confirmation(1);
    
                storage = await inventoryInstance.storage();
            });
    
            it("Then assigns the new item to the user", async () => {
                const obj = await getInventoryItemAt(storage, "user_123", 3, 1);
    
                expect(obj).toEqual({
                    data: {
                        XP: "97"
                    }
                });
            });
        });
    });
});


src/warehouse.ts
import {
    ContractAbstraction,
    ContractMethod,
    ContractProvider,
    MichelsonMap
} from "@taquito/taquito";

import { BigNumber } from "bignumber.js";

export interface MichelsonWarehouseItem {
    available_quantity: BigNumber;
    data: MichelsonMap<string, string>;
    frozen: boolean;
    item_id: BigNumber;
    name: string;
    total_quantity: BigNumber;
    [key: string]: MichelsonMap<string, string> | boolean | string | BigNumber;
}

export interface DataField {
    [k: string]: string;
}

export interface InstancesKey {
    item_id: number;
    instance_number: number;
}

export type ItemsBigMap = MichelsonMap<string, MichelsonWarehouseItem>;
export type InstancesBigMap = MichelsonMap<
    InstancesKey,
    MichelsonWarehouseItem
>;

export interface WarehouseStorage {
    owners: string[];
    version: string;
    items: ItemsBigMap;
    instances: InstancesBigMap;
}

export interface WarehouseContract
    extends ContractAbstraction<ContractProvider> {
    storage: <WarehouseStorage>() => Promise<WarehouseStorage>;
    methods: {
        add_item(
            available_quantity: number,
            data: MichelsonMap<string, string>,
            frozen: boolean,
            item_id: number,
            name: string,
            total_quantity: number
        ): ContractMethod<ContractProvider>;

        update_item(
            available_quantity: number,
            data: MichelsonMap<string, string>,
            frozen: boolean,
            item_id: number,
            name: string,
            total_quantity: number
        ): ContractMethod<ContractProvider>;

        freeze_item(item_id: number): ContractMethod<ContractProvider>;

        assign_item(
            item_id: number,
            instance_number: number,
            user_id: string
        ): ContractMethod<ContractProvider>;

        update_instance(
            item_id: number,
            instance_number: number,
            data: MichelsonMap<string, string>
        ): ContractMethod<ContractProvider>;

        transfer_instance(
            item_id: number,
            instance_number: number,
            user_id: string
        ): ContractMethod<ContractProvider>;

        add_owner(address: string): ContractMethod<ContractProvider>;

        remove_owner(address: string): ContractMethod<ContractProvider>;
    };
}

export interface JSONWarehouseItem {
    available_quantity: number;
    data: { [k: string]: string };
    frozen: boolean;
    item_id: number;
    name: string;
    total_quantity: number;
    [key: string]: { [k: string]: string } | boolean | string | number;
}

export type LinearWarehouseItem = [
    number,
    MichelsonMap<string, string>,
    boolean,
    number,
    string,
    number
];

export class WarehouseItem {
    readonly available_quantity: BigNumber;
    readonly data: DataField;
    readonly frozen: boolean;
    readonly item_id: BigNumber;
    readonly name: string;
    readonly total_quantity: BigNumber;

    constructor(object: { [k: string]: unknown }) {
        this.available_quantity = getKey(
            object,
            "available_quantity"
        ) as BigNumber;
        this.data = getKey(object, "data") as DataField;
        this.frozen = getKey(object, "frozen") as boolean;
        this.item_id = getKey(object, "item_id") as BigNumber;
        this.name = object.name as string;
        this.total_quantity = getKey(object, "total_quantity") as BigNumber;

        this.validateData(this.data);
    }

    toMichelsonArguments(): LinearWarehouseItem {
        const warehouseItem = {
            available_quantity: this.available_quantity,
            data: MichelsonMap.fromLiteral(this.data),
            frozen: this.frozen,
            item_id: this.item_id,
            name: this.name,
            total_quantity: this.total_quantity
        } as MichelsonWarehouseItem;

        return Object.keys(warehouseItem)
            .sort()
            .map((key: string) => warehouseItem[key]) as LinearWarehouseItem;
    }

    static fromMichelson(michelson: MichelsonWarehouseItem): JSONWarehouseItem {
        return {
            available_quantity: michelson.available_quantity.toNumber(),
            data: Object.fromEntries(michelson.data.entries()),
            frozen: michelson.frozen,
            item_id: michelson.item_id.toNumber(),
            name: michelson.name.toString(),
            total_quantity: michelson.total_quantity.toNumber()
        };
    }

    private validateData(data: DataField): void {
        if (Object.values(data).some((datum) => typeof datum !== "string")) {
            throw new Error(`WarehouseItem: Data must be 'string'`);
        }
    }
}

function getKey(object: { [k: string]: unknown }, key: string) {
    if (!(key in object)) {
        throw new Error(
            `WarehouseItem: Key ${key} is not present in warehouseItem`
        );
    } else {
        return object[key];
    }
}


src/inventoryItem.ts
import {
    ContractAbstraction,
    ContractMethod,
    ContractProvider,
    MichelsonMap
} from "@taquito/taquito";

export type JSONInventoryItem = { [k: string]: string };

export type InventoryStorage = MichelsonMap<
    string,
    MichelsonMap<string, MichelsonMap<string, MichelsonMap<string, string>>>
>;

export interface InventoryContract
    extends ContractAbstraction<ContractProvider> {
    storage: <InventoryStorage>() => Promise<InventoryStorage>;
    methods: {
        assign_item(
            data: MichelsonMap<string, string>,
            instance_number: number,
            item_id: number,
            user_id: string
        ): ContractMethod<ContractProvider>;

        update_item(
            data: MichelsonMap<string, string>,
            instance_number: number,
            item_id: number,
            user_id: string
        ): ContractMethod<ContractProvider>;
    };
}

export function toMichelsonInventoryItem(
    data: JSONInventoryItem
): MichelsonMap<string, string> {
    return MichelsonMap.fromLiteral(data) as MichelsonMap<string, string>;
}

export function fromMichelsonInventoryItem(
    data: MichelsonMap<string, string>
): JSONInventoryItem {
    return Object.fromEntries(data.entries());
}


src/warehouseItem.ts
import {
    ContractAbstraction,
    ContractMethod,
    ContractProvider,
    MichelsonMap
} from "@taquito/taquito";

import { BigNumber } from "bignumber.js";

export interface MichelsonWarehouseItem {
    available_quantity: BigNumber;
    data: MichelsonMap<string, string>;
    frozen: boolean;
    item_id: BigNumber;
    name: string;
    total_quantity: BigNumber;
    [key: string]:
        | MichelsonMap<string, string>
        | boolean
        | string
        | BigNumber;
}

export type WarehouseBigMap = MichelsonMap<string, MichelsonWarehouseItem>;

export interface WarehouseStorage {
    owner: string;
    version: string;
    warehouse: WarehouseBigMap;
}

export interface WarehouseContract
    extends ContractAbstraction<ContractProvider> {
    storage: <WarehouseStorage>() => Promise<WarehouseStorage>;
    methods: {
        add_item(
            available_quantity: number,
            data: MichelsonMap<string, string>,
            frozen: boolean,
            item_id: number,
            name: string,
            total_quantity: number
        ): ContractMethod<ContractProvider>;

        update_item(
            available_quantity: number,
            data: MichelsonMap<string, string>,
            frozen: boolean,
            item_id: number,
            name: string,
            no_update_after: string | undefined,
            total_quantity: number
        ): ContractMethod<ContractProvider>;

        freeze_item(item_id: number): ContractMethod<ContractProvider>;

        assign_item_proxy(
            inventory_address: string,
            item_id: number,
            instance_number: number
        ): ContractMethod<ContractProvider>;
    };
}

export interface WarehouseData {
    [k: string]: string;
}

export interface JSONWarehouseItem {
    available_quantity: number;
    data: { [k: string]: string };
    frozen: boolean;
    item_id: number;
    name: string;
    total_quantity: number;
    [key: string]: 
        | { [k: string]: string }
        | boolean
        | string
        | number;
}

export type LinearWarehouseItem = [
    number,
    MichelsonMap<string, string>,
    boolean,
    number,
    string,
    number
];

export class WarehouseItem {
    readonly available_quantity: BigNumber;
    readonly data: WarehouseData;
    readonly frozen: boolean;
    readonly item_id: BigNumber;
    readonly name: string;
    readonly total_quantity: BigNumber;

    constructor(object: { [k: string]: unknown }) {
        this.available_quantity = getKey(
            object,
            "available_quantity"
        ) as BigNumber;
        this.data = getKey(object, "data") as WarehouseData;
        this.frozen = getKey(object, "frozen") as boolean;
        this.item_id = getKey(object, "item_id") as BigNumber;
        this.name = object.name as string;
        this.total_quantity = getKey(object, "total_quantity") as BigNumber;

        this.validateData(this.data);
    }

    toMichelsonArguments(): LinearWarehouseItem {
        const warehouseItem = {
            available_quantity: this.available_quantity,
            data: MichelsonMap.fromLiteral(this.data),
            frozen: this.frozen,
            item_id: this.item_id,
            name: this.name,
            total_quantity: this.total_quantity
        } as MichelsonWarehouseItem;

        return Object.keys(warehouseItem)
            .sort()
            .map((key: string) => warehouseItem[key]) as LinearWarehouseItem;
    }

    static fromMichelson(michelson: MichelsonWarehouseItem): JSONWarehouseItem {
        return {
            available_quantity: michelson.available_quantity.toNumber(),
            data: Object.fromEntries(michelson.data.entries()),
            frozen: michelson.frozen,
            item_id: michelson.item_id.toNumber(),
            name: michelson.name.toString(),
            total_quantity: michelson.total_quantity.toNumber()
        };
    }

    private validateData(data: WarehouseData): void {
        if (Object.values(data).some((datum) => typeof datum !== "string")) {
            throw new Error(`WarehouseItem: Data must be 'string'`);
        }
    }
}

function getKey(object: { [k: string]: unknown }, key: string) {
    if (!(key in object)) {
        throw new Error(
            `WarehouseItem: Key ${key} is not present in warehouseItem`
        );
    } else {
        return object[key];
    }
}


misc/operation_confirmation.js
const { TezosToolkit, MichelsonMap } = require("@taquito/taquito");
const { InMemorySigner } = require("@taquito/signer");

const warehouseContract = require("../build/contracts/warehouse.json");

const Tezos = new TezosToolkit("http://localhost:20000");

Tezos.setProvider({
    signer: new InMemorySigner(
        "edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq"
    )
});

const context = {};

async function originateWarehouse() {
    const originationOp = await Tezos.contract.originate({
        code: warehouseContract,
        storage: {
            owners: ["tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb"],
            version: "1",
            items: MichelsonMap.fromLiteral({}),
            instances: MichelsonMap.fromLiteral({})
        }
    });

    context.warehouseAddress = originationOp.contractAddress;

    await originationOp.confirmation();
}

async function addItem(...item) {
    const contract = await Tezos.contract.at(context.warehouseAddress);

    return contract.methods.add_item(...item).send();
}

async function updateItem(...item) {
    const contract = await Tezos.contract.at(context.warehouseAddress);

    const originationOp = await contract.methods.update_item(...item).send();

    await originationOp.confirmation();
}

(async function run() {
    try {
        console.log("originating warehouse", new Date());
        await originateWarehouse();

        console.log(`originated at`, context.warehouseAddress, new Date());

        const operation = await addItem(
            10,
            MichelsonMap.fromLiteral({
                XP: "97"
            }),
            false,
            0,
            "Karim Benzema",
            10
        );

        console.log(JSON.stringify(operation.hash));

        // console.log('item generated')

        // console.log('counter', (await Tezos.rpc.getContract(signer)), new Date());

        // await addItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     0,
        //     "Karim Benzema",
        //     10
        // )

        // console.log('counter', (await Tezos.rpc.getContract(signer)), new Date());

        // await updateItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     0,
        //     "Karim Benzema",
        //     10
        // )

        // console.log("generated item 1", new Date());
        // console.log('counter', (await Tezos.rpc.getContract(signer)));

        // await addItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     1,
        //     "Karim Benzema",
        //     10
        // )
        // console.log('counter', (await Tezos.rpc.getContract(signer)));

        // await updateItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     1,
        //     "Karim Benzema",
        //     10
        // )
        // console.log('counter', (await Tezos.rpc.getContract(context.warehouseAddress)));

        // console.log("generated item 2", new Date());

        // await addItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     2,
        //     "Karim Benzema",
        //     10
        // )

        // await updateItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     2,
        //     "Karim Benzema",
        //     10
        // )

        // console.log("generated item 3", new Date());

        // await addItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     3,
        //     "Karim Benzema",
        //     10
        // )

        // await updateItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     3,
        //     "Karim Benzema",
        //     10
        // )

        // console.log("generated item 4", new Date());

        // await addItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     4,
        //     "Karim Benzema",
        //     10
        // )

        // await updateItem(
        //     10,
        //     MichelsonMap.fromLiteral({
        //         XP: "97"
        //     }),
        //     false,
        //     4,
        //     "Karim Benzema",
        //     10
        // )

        // console.log("generated item 5", new Date());
    } catch (err) {
        console.log(err);
    }
})();
